### PR TITLE
Add or remove participation time to single user in online contest

### PR DESCRIPTION
**Pull request by gogo4ds**
Implement functionality to add contest Compete time by user's username
Closes https://github.com/SoftUni-Internal/suls-issues/issues/4734

### CODE REVIEW

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 1]
```diff
@@ -375,43 +368,79 @@ public ActionResult ChangeActiveParticipantsEndTime(ChangeTimeForParticipantsVie
...
-            this.participantsBusiness
-                .UpdateContestEndTimeForAllParticipantsByContestByParticipantContestStartTimeRangeAndTimeIntervalInMinutes(
-                    model.ContesId,
-                    model.TimeInMinutes,
-                    model.ParticipantsCreatedAfterDateTime.Value,
-                    model.ParticipantsCreatedBeforeDateTime.Value);
+            ServiceResult result;
 
-            var minutesForDisplay = model.TimeInMinutes.ToString();
+            if (model.ChangeByTimeInterval)
```
Comment (achobanov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: Does this have to be a single `Action`? As is it effectively combines two different methods, using a `flag` and `if statement` to differentiate between the required action and execute the appropriate logic with the only benefit I see is to call common error-handling only once. I think it would be cleaner if it was split in two actual methods.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 2]
```diff
@@ -375,43 +368,79 @@ public ActionResult ChangeActiveParticipantsEndTime(ChangeTimeForParticipantsVie
...
+
+                this.TempData.AddInfoMessage(systemMessage);
+            }
+            else
+            {
+                this.TempData.AddDangerMessage(result.Error);
             }
 
-            this.TempData.AddInfoMessage(sb.ToString());
+            void PrepareViewBagData() => this.ViewBag.CurrentUsername = this.UserProfile.UserName;
```
Comment (achobanov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: What is that syntax? I haven't seen it before. Is this a short-hand for a private method? If so why not define it in the controller body, as opposed to inside the method body. Interesting.

File: Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeTimeForParticipantsViewModel.cs [hunk 3]
```diff
@@ -37,11 +37,18 @@ public class ChangeTimeForParticipantsViewModel
...
         public DateTime? ParticipantsCreatedAfterDateTime { get; set; }
 
         [Display(Name = "Participants_created_before", ResourceType = typeof(Resource))]
-        [Required(ErrorMessageResourceName = "Time_required_error", ErrorMessageResourceType = typeof(Resource))]
         public DateTime? ParticipantsCreatedBeforeDateTime { get; set; }
+
+        [UIHint("UsersSimpleComboBox")]
+        public string UserId { get; set; }
+
+        [Display(Name = "Change_time_by_time_interval", ResourceType = typeof(Resource))]
```
Comment (achobanov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeTimeForParticipantsViewModel.cs]: Don't you think it is more robust to use `nameof(Resource,Change_time_by_time_interval)`?

File: Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeTimeForParticipantsViewModel.cs [hunk 4]
```diff
@@ -37,11 +37,18 @@ public class ChangeTimeForParticipantsViewModel
...
+        public bool ChangeByTimeInterval { get; set; } = true;
+
+        [Display(Name = "Change_time_by_user", ResourceType = typeof(Resource))]
```
Comment (achobanov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeTimeForParticipantsViewModel.cs]: Same as previous - `nameof`

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 5]
```diff
@@ -178,17 +212,40 @@
...
+
             function changeAddSubractButton(value) {
                 var addButton = $(addButtonSelector);
                 var subtractButton = $(subtractButtonSelector);
-                var hiddenClass = "hidden";
 
                 if (value >= 0) {
-                    addButton.removeClass(hiddenClass);
-                    subtractButton.addClass(hiddenClass);
+                    addButton.removeClass('hidden');
```
Comment (achobanov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Why did you remove the variable?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs (hunk 2 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: This ViewBag data is specific for this action and it has to be set every time before returning the view, which is done several times in the body of the method and a [local function](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions) seems like an appropriate solution for not repeating the code.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml (hunk 5 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: because this is the only place where a variable is used like this, and I think is more readable to not use a variable for html property or class

Comment (achobanov) [reply Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: I would use the variable still, but its up to preference, I guess.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 6]
```diff
@@ -356,62 +357,86 @@ public ActionResult ChangeActiveParticipantsEndTime(ChangeTimeForParticipantsVie
...
 
-            var notUpdatedUsersUsernames = this.participantsBusiness
-                .GetAllParticipantsWhoWouldBeReducedBelowDefaultContestDuration(
-                    model.ContesId,
-                    model.TimeInMinutes,
-                    model.ParticipantsCreatedAfterDateTime.Value,
-                    model.ParticipantsCreatedBeforeDateTime.Value)
-                .Select(u => u.User.UserName)
-                .ToList();
+                //var participant = this.participantsData
```
Comment (achobanov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: Do you have to keep those comments?

File: Open Judge System/OJS.Services/Business/Participants/ParticipantsBusinessService.cs [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
+
     using OJS.Data.Models;
+    using OJS.Services.Common;
     using OJS.Services.Data.Contests;
     using OJS.Services.Data.Participants;
 
     public class ParticipantsBusinessService : IParticipantsBusinessService
     {
+        private const string ContestDoesNotExistErrorMessage = "Contest does not exist!";
+        private const string ContestDoesNotHaveDurationErrorMessage = "The contest does not have duration set!";
```
Comment (jicata) [inline Open Judge System/OJS.Services/Business/Participants/ParticipantsBusinessService.cs]: Shouldn't we aim to localize this kind of messages?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 8]
```diff
@@ -577,5 +609,60 @@ private void AddProblemGroupsToContest(Contest contest, int problemGroupsCount)
...
+        private string GetMessageForChangedParticipantsEndTimeByTimeInterval(
+            int timeInMinutes,
+            string contestName,
+            ICollection<string> notUpdatedParticipantUsernames)
+        {
+            var minutesForDisplay = timeInMinutes.ToString();
+
+            var sb = new StringBuilder();
+
+            var message = timeInMinutes >= 0
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: IMO an if statement is better here:
```
var formatString = Resource.Added_time_to_participants_online;
if (timeInMinutes < 0) {
  formatString = Resource.Added_time_to_participants_online;
  minutesForDisplay= minutesForDisplay.Substring(1, minutesForDisplay.Length - 1);
}
var message = string.format(formatString, minutesForDisplay, contestname);
```

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 9]
```diff
@@ -577,5 +609,60 @@ private void AddProblemGroupsToContest(Contest contest, int problemGroupsCount)
...
         }
+
+        private string GetMessageForChangedParticipantsEndTimeByTimeInterval(
+            int timeInMinutes,
+            string contestName,
+            ICollection<string> notUpdatedParticipantUsernames)
+        {
+            var minutesForDisplay = timeInMinutes.ToString();
+
+            var sb = new StringBuilder();
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: Using StringBuilder implies a lot of string concatenation, and this is not such a case. Simple string is better IMO.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs [hunk 10]
```diff
@@ -577,5 +609,60 @@ private void AddProblemGroupsToContest(Contest contest, int problemGroupsCount)
...
+                ? string.Format(
+                    Resource.Added_time_to_single_participant_online,
+                    minutesForDisplay,
+                    username,
+                    contestName)
+                : string.Format(
+                    Resource.Subtracted_time_from_single_participant_online,
+                    minutesForDisplay.Substring(1, minutesForDisplay.Length - 1),
+                    username,
+                    contestName);
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs]: same as above

File: Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeParticipationEndTimeViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    using OJS.Data.Models;
+
+    using Resource = Resources.Areas.Administration.Contests.Views.ChangeTime;
+
+    public class ChangeParticipationEndTimeViewModel
+    {
+        private static readonly int DefaultBufferTimeInMinutes = 30;
+
+        public ChangeParticipationEndTimeViewModel()
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeParticipationEndTimeViewModel.cs]: IMO, when empty default constructor is neccessary,best to comment it with something like " // Default constructor needed for REASON". Otherwise it is just wierd.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 12]
```diff
@@ -109,86 +81,121 @@
     <script>
         ($(function () {
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Are the outer brackets neccesarry?
```
$(function() {
...);
```
works just fine

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 13]
```diff
@@ -109,86 +81,121 @@
...
             var timeInput = $('#TimeInMinutes');
+            var addButtonSelector = '#add-time-for-participants-btn';
+            var subtractButtonSelector = '#subtract-time-for-participants-btn';
+            var buttons = [addButtonSelector, subtractButtonSelector];
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: `buttonSelectors`?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 14]
```diff
@@ -109,86 +81,121 @@
...
 
-            timeInput.on('input',
-                function () {
-                    changeAddSubractButton(timeInput.val());
-                });
+            timeInput.on('input', function () {
+                changeAddSubractButton(timeInput.val());
+            });
+
+            $(buttons.join()).click(function () {
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Best to make it `.join(' ')`, i.e. put explicit whitespace as parameter. Some browsers, the older ones,  jon without a whitespace.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 15]
```diff
@@ -109,86 +81,121 @@
...
+                minutes.text(Math.abs(inputValue));
+
+                $('#username-confirmation-input').on('input',
+                    function () {
+                        var value = $(this).val();
+                        if (value === '@this.User.Identity.Name') {
+                            submitFormButton.attr('disabled', false);
+                        } else {
+                            submitFormButton.attr('disabled', true);
+                        }
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]:
```
var submitFormButtonState = value === '@this.User.Identity.Name';
submitFormButton.attr('disabled', submitFormButtonState);
```

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 16]
```diff
@@ -109,86 +81,121 @@
...
+                }
 
-            var addButtonSelector = '#add-time-for-participants-btn';
-            var subtractButtonSelector = '#subtract-time-for-participants-btn';
+                createdBefore.text(createdBeforeValue);
+                createdAfter.text(createdAfterValue);
+                usernamePlaceholder.text(participantUsername);
+                minutes.text(Math.abs(inputValue));
+
+                $('#username-confirmation-input').on('input',
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Is this event handler attached inside the button click handler on purpose?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 17]
```diff
@@ -109,86 +81,121 @@
...
+                            submitFormButton.attr('disabled', false);
+                        } else {
+                            submitFormButton.attr('disabled', true);
+                        }
+                    });
+
+                confirmWindow.show();
+                changeTimeContainer.css('opacity', 0.3);
+
+                submitFormButton.click(
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Is this event handler attached inside the button click handler on purpose?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 18]
```diff
@@ -109,86 +81,121 @@
...
+                        $('form input[name=TimeInMinutes]').val(timeInput.val());
+  
+                        if ($('#change-time-by-interval').is(':visible')) {
+                            $('#change-time-by-interval-form').submit();
+                        } else if ($('#change-time-by-user').is(':visible')) {
+                            $('#change-time-by-user-form').submit();
+                        }
+                    });
+
+                $('#cancel-btn-confirm-pop-up').click(
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Is this event handler attached inside the button click handler on purpose?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 19]
```diff
@@ -109,86 +81,121 @@
...
+                    function () {
+                        confirmWindow.hide();
+                        changeTimeContainer.css('opacity', 1);
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Can this be done done with CSS class instead?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 20]
```diff
@@ -109,86 +81,121 @@
...
+                    });
+            });
+
+            function setConfirmWindowHeader() {
+                var confirmWindowHeader = $('#confirm-window-header');
+                var headerPrefix = '@Resource.Confirmation_message_header_prefix ';
+                var headerText = '';
+
+                if ($('#change-time-by-interval').is(':visible')) {
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: If the above is done with a CSS class, this check can be done less hacky

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 21]
```diff
@@ -109,86 +81,121 @@
...
+                    headerText = '@Resource.Confirmation_message_header_change_by_interval';
+                } else if ($('#change-time-by-user').is(':visible')) {
+                    headerText = '@Resource.Confirmation_message_header_change_by_user';
+                }
 
-            var buttons = [addButtonSelector, subtractButtonSelector];
+                var parser = new DOMParser;
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: `var parser = new DOMParser();`

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml (hunk 5 above)
Comment (Minkov) [reply Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: Using the string is more expressive here. `hidden` is common enough class`

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml [hunk 22]
```diff
@@ -109,86 +81,121 @@
...
-                    addButton.removeClass(hiddenClass);
-                    subtractButton.addClass(hiddenClass);
+                    addButton.removeClass('hidden');
+                    subtractButton.addClass('hidden');
                 } else if (value < 0) {
-                    addButton.addClass(hiddenClass);
-                    subtractButton.removeClass(hiddenClass);
+                    addButton.addClass('hidden');
+                    subtractButton.removeClass('hidden');
+                }
```
Comment (Minkov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: This can be simplified with built-in function and no if-els:
```
addButton.toggleCalss('hidden');
subtractButton.toggleClass('hidden');
```

File: Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeParticipationEndTimeViewModel.cs (hunk 11 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Administration/ViewModels/Contest/ChangeParticipationEndTimeViewModel.cs]: I think in this case it is not needed, since it's usage location can be easily tracked in the same file and if removed, a build error emerges immediately.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml (hunk 20 above)
Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Administration/Views/Contests/ChangeActiveParticipantsEndTime.cshtml]: The above CSS class is for another container
